pymedphys/_dev/propagate.py

```diff
@@ -0,0 +1,104 @@
+# Copyright (C) 2020 Simon Biggs
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import pathlib
+import subprocess
+import textwrap
+
+from pymedphys._imports import black, tomlkit
+
+REPO_ROOT = pathlib.Path(__file__).resolve().parent.parent.parent
+PYPROJECT_TOML_PATH = REPO_ROOT.joinpath("pyproject.toml")
+
+LIBRARY_PATH = REPO_ROOT.joinpath("pymedphys")
+VERSION_PATH = LIBRARY_PATH.joinpath("_version.py")
+
+
+def propagate_all(_):
+    propagate_version()
+    propagate_extras()
+    propagate_requirements()
+
+
+def read_pyproject():
+    with open(PYPROJECT_TOML_PATH) as f:
+        pyproject_contents = tomlkit.loads(f.read())
+
+    return pyproject_contents
+
+
+def propagate_version():
+    pyproject_contents = read_pyproject()
+
+    version_string = deps = pyproject_contents["tool"]["poetry"]["version"]
+    version_list = version_string.split(".")
+
+    for i, item in enumerate(version_list):
+        try:
+            version_list[i] = int(item)
+        except:
+            pass
+
+    version_contents = textwrap.dedent(
+        f"""\
+        # DO NOT EDIT
+        # This file is autogenerated by `pymedphys dev propagate`
+
+        version_info = {version_list}
+        __version__ = "{version_string}"
+        """
+    )
+
+    version_contents = black.format_str(version_contents, mode=black.FileMode())
+
+    with open(VERSION_PATH, "w") as f:
+        f.write(version_contents)
+
+
+def propagate_requirements():
+    subprocess.check_call(
+        "poetry export -E dev -f requirements.txt --output requirements.txt", shell=True
```

Comment:
dev is probably the most complete, but is requirements.txt then going to be based on [dev] ?

Reply:
Yeah, that'll mean that the requirements.txt file produced will be equivalent to a dev install. I can see why it may be better to opt for user instead...

Reply:
I think user would be better.
If you are a pymedphys developer, you are going to use poetry (you pretty much have to) and not use the requirements.txt
But if you are a user, you'd really be a lot happier to just use the requirements.txt (and presumably pip would use that to address installing dependencies while it's installing pymedphys?)

Reply:
Yup, that makes sense. To install with pip a setup.py is also needed. I'll use add a setup.py creation to this propagate CLI also:

https://github.com/python-poetry/poetry/issues/761#issuecomment-521124268

Reply:
In the end the setup.py produced by dephell was nasty. So I extracted poetry's version instead.